### ISSUE TITLE
[Feature Request] Sessions

### ISSUE THREAD

**Issue by frankcorneliusmartin** [vantage6/vantage6#943]
## Advantages
* Unified workflow for the user regardless of which type of data source is being used (CSV, SQL, SPARQL, ...). For example because we have a tabular data frame we can share the variables with the user to select for the analysis (dropdown in the UI)
* Metadata can be shared prior the compute steps (e.g. column names, so the user can select them for a dropdown)
* Users are able to share sessions with other users (can be blocked with permissions)
* Obtaining data from the source only happens once per sessions. Improving speed especially when querying requires a lot of time. 
* User is able to combine different functions from different containers 
* Code is reusable and avoids having many copies of preprocessing tools. This helps keeping overview of our code and improves the code quality
* Data extraction / pre-processing and compute step can be written in a completely different langauge
* Users with sufficient privileges can prepare sessions for other users to use. For example if you have a collaboration administrator who creates a Study #812 can also create a session and preparing a dataset that can be used in the study. Other users who cannot create new sessions can then reuse these
* Containers are only mounted with the files and provided with environment variables they really need
* Validate that the intermediate dataframes have the same column names before the analysis (compute) is started. Making it more user-friendly as these messages can be shown in the UI.

### Server

- [x] Extend database with `sessions` model -> added `Session`, `NodeSession` and `NodeSessionConfig`
  - [x] Assign a `session_id` to a each task and to a set of organizations or study (within the collaboration)
  - [x] A session consists of a UUID and a label. UUID needs to be server wide unique and the label needs to be collaboration unique
  - [x] Unittests
- [x] Add endpoints to the vantage6 server to manage sessions; 
  - [x] add additional permissions (user can create,read and delete sessions, nodes can read them)
  - [x] create 
  - [x] read
  - [x] delete
  - [x] unittests
- [x] Session initialization
  - [x] Add session pipelines to POST body of the `/session` endpoint
  - [x] Support multiple local pipelines
  - [x] Local pipelines should be executed in serial
  - [x] Add `depends_on` and `type` columns to the task model
- [ ] Extend vantage6 server
  - [x] When task is created it should be part of a session
  - [ ] add option to the collaboration to chain only tasks from the same image
  - [x] Loosen Container token / with option to not do this
  - [x]  Extend /session endpoints to contain actions (list of tasks)
- [x] Add endpoints for nodes to patch their status
- [x] Remove session 

### Node 
- [x] Add socket events `session`
- [x] Replace Temporary volumes met Session volumes
- [ ] Update node_session state on the server from the node / algorithm container (algorithm tools)
- [x] Intermediate results should be deleted
- [ ] Make a clean session trigger
- [x] Session triggers should be executed in parallel (as we can only do preprocessing after data extraction)
- [x] Handle different actions (types of containers)

### UI
- [x] Add Session step in the create tasks section
- [x] Add Session overview step 
- [ ] Create sessions with `once-per-session` steps so that these can be shared with user that are not allowed to create sessions

### Python Client
- [x] Make sessions available
  - [x] `client.session.create`
  - [x] `session.pipeline.create`
  - [ ] ...

### Algorithm Store
- [ ] Add action meta data to functions

**Comment by jvsoest** [comment]
How would this relate to the app store? Does every container represent a separate app (with pre-requisite apps)? Or is an "app" a collection of containers?

Basically this complicates what we need to achieve in strong-aya, depending on how we see what an "app" represents.
@lsago something to consider as well together with @bartvanb

**Comment by bartvanb** [comment]
Good question @jvsoest. I think the easiest way to implement this is that the algorithm store treats each container as separate algorithm, and there is a 'chaining' configuration for each algorithm (i.e. which algorithms are allowed to follow/precede it? Should it always be the first/last?). This may then later be replaced by showing the relevant algorithm combinations if that makes more sense. I can also imagine we create an algorithm store setting where you can choose between the two.

Another idea that is relevant here are template tasks (see #857), which are basically blueprints for tasks where only a few variables have still to be defined. These can be run from a UI / Python client as 'Quick task' where researcher only have to fill in a few things. We should make sure that we can create templates for chained tasks as well.

**Comment by frankcorneliusmartin** [comment]
Best would be to implement #154 first. As then we do not have to rely on the temporary volumes (which are scheduled for removal: #248) and run ids.

**Comment by frankcorneliusmartin** [comment]
I think we should also include some hooks so that data station owners can define mandatory intermediate containers: 

```mermaid
graph LR
    subgraph DataExtraction
        A[step]
    end
    subgraph Preprocessing
        C[step 1]
        D[step n]
        C -.-> D
    end
    subgraph Algorithm
        E[Algorithm]
    end 
    subgraph Postprocessing
        F[step 1]
        G[step n]
        F -.-> G
    end

A -->|hook1| C
D -->|hook2| E
E -->|hook3| F
```

**Comment by bartvanb** [comment]
Nice to have a more detailed plan! Mostly sounds good, a few comments/questions:
- How would the 'hooks' work?
- The `/output` directory is normally for algorithm results right? How will we make absolutely sure that if there is a dataframe there, that this is never ever sent to the central server?
- Is #154 now something that you plan to implement as well when working on this?
- Handover: I like the idea of keeping track of steps in a file. Maybe there can be a file per task (e.g. `intermeditate_steps_task_X.parquet`) that details the steps for a single task, so that we prevent trying to read/write from multiple tasks to the same file simultaneously

**Comment by frankcorneliusmartin** [comment]
Some additional thoughts:

* We should also consider how to deal with multiple datasets
* Is parquet supported in the most common used data science languages (Python, R, Julia, Java, ?)

> Is https://github.com/vantage6/vantage6/issues/154 now something that you plan to implement as well when working on this?

I think I use the temporary volumes for now. I am not sure but I think @lsago was working on the file storage?

**Comment by lsago** [comment]
> I think I use the temporary volumes for now. I am not sure but I think @lsago was working on the file storage?
Yes, #154 is something I started to work on, although have neglected recently... Hopefully I can pick it up again soon. Still something I very much think is worth doing.

**Comment by frankcorneliusmartin** [comment]
# WIP

The user is able to define `data-extraction`, `preprocessing`, `compute` and `post-processing` actions in steps in a `pipeline`. A single step can also execute a combination of these actions (*legacy* algorithm containers have all of them). And not every pipeline should have all components. The only mandatory action is `compute` (there can be a container which computes something without the need of the local data). 

The order of the actions is always the same:

1. data-extraction
2. preprocessing
3. compute
4. post-processing

Actions are stored as functions in images. A single image can, but does not have to, contain have all these steps as separate functions in them. It is also possible to use different functions from different images.

# User input / Python Client
A first attempt:
```json
"organizations": 
[
  {
    "id": 1,
    "pipeline": [
       {
          "image": "data-extraction-1", 
          "function": "some-action-function-in-this-image",
          "args": ["arg1", "arg2"],
          "kwargs": {"a": 1, "b": 2}
       }, 
       ... 
    ]
  } 
]
```

# Node 
The node should execute the steps one-by-one. The node should be able to store the **RunOncePerAnalysis** resulting `df` somewhere the successor (compute) containers can access them. This could be the `TemporaryFolder` however we should allow some freedom in in linking the temporary volumes to the algorithm containers. These are tied to `run_ids  now, but the if the containers come in different batches these will have different run_ids. We could consider sharing a secret with the client in order to connect to the correct tmp folder.


## Local Handover
* step id, so that the container can find the latest step
* output type of each step, so that the next container can check if there is an input it can use
* which image was used (optionally which input was used?)
* Where the output is located
* metadata like timestamp of start, end

The file could contain something like: 

| action | type | file | metadata |
|--------|--------|--------|--------|
| 1 |  `data-extraction` | `/data/extracted_1` | {} |
| 2 | `preprocessing` | `data/preprocessed_1` | {} |
| 3 | `compute` | `data/result` | {} |

## Provisioning Algorithm Containers
Each action requires different data from the node. Currently algorithm containers are provided with a set of environment vars and mounts: `INPUT_FILE`, `OUTPUT_FILE`, `TOKEN_FILE`, `TEMPORARY_FOLDER`, `HOST`, `PORT`, `API_PATH`, `HTTP_PROXY`, `[LABEL]_DATABASE_URI`, `[LABEL]_DATABASE_TYPE`, `DB_LABELS`, `[LABEL]_DB_PARAM_[KEY]` `USER_REQUESTED_DATABASE_LABELS`, `NO_PROXY`.

# Algorithm store
Each function should have an action tied to it.

# Algorithm Tools
These should be extended to contain additional decorators for the different actions. Each action should gets its own input and environment vars that could/should be accessed. 

```python
@action("compute")  # this is the new decorator
@data(1)
@client
```

This decorator can control what happens with the IO. Thus writing to the `TemporaryFolder` or write it to the `output` file, it could potentially also provide arguments to the functions.

# UI 
The UI should be extended to allow to build pipelines. There should be a separate (optional) section for **OncePerAnalysis** where they can define preprocessing steps which only have to be run once. Then the preprocessing should already be run so that the column(=variable) names can be retrieved to be used in the *compute step*.

I am a little bit overwhelmed by the amount of options we have to bring this feature to the user through the UI. @bartvanb maybe have a whiteboard brainstorm session about this?

# Random Thoughts / Notes
* The last container in the pipeline should always be the container that returns data to the central server. This container should write to the `output` file. This can either be the `compute` or the `post-processing` action. 
* Can we prevent people from using other types of containers for returning data. E.g. we should prevent that only a data-extraction container is send. Or do we solely rely on the reviewing process of these containers?
* Should the users be able to define pipelines at all? Or should there be a single pipeline per node?
* This structure does not allow for grouping data-extraction and preprocessing to be run once per run_id. We could add an additional key, something like `run_once: true`?
* Should we have a seperate type for aggregator/orchastrator?
* ....


> e.g. intermeditate_steps_task_X.parquet) that details the steps for a single task, so that we prevent trying to read/write from multiple tasks to the same file simultaneously

Good point. Let's us forget about the format for now. And see what needs to be stored:

**Comment by dsmits** [comment]
I have a couple of thoughts:

Since this is basically about running workflows, have you considered if workflows systems like [cwl](https://www.commonwl.org/) fit into this? This chaining of operations also makes me think of my colleague lourens' [mahiru](https://github.com/SecConNet/mahiru) where a cwl-like workflow system is used to chain tasks together. This is combined with policies on where things are allowed to run. Might be too hard to integrate in v6 but I just wanted to mention it.

I guess you're going for a different direction though, with the session idea. I do see potential for an improved UX at the python client. You could have local placeholder variables for the outputs that you can apply other functions to that will then be executed remotely.

Another thing, will the central algorithms still only be able to execute partial functions that are within the same image? I think this will still hold but I want to check. I think it is a good restriction to have.

Lastly, just to be sure, I think the main point of this idea is that you don't have to repeat preprocessing steps right? You're not chaining multiple compute tasks one after the other I think.

**Comment by frankcorneliusmartin** [comment]
> Since this is basically about running workflows, have you considered if workflows systems like [cwl](https://www.commonwl.org/) fit into this? This chaining of operations also makes me think of my colleague lourens' [mahiru](https://github.com/SecConNet/mahiru) where a cwl-like workflow system is used to chain tasks together. This is combined with policies on where things are allowed to run. Might be too hard to integrate in v6 but I just wanted to mention it.

I will have a look at these, although these seem to be a bit more focused on chaining command line applications. 

> I guess you're going for a different direction though, with the session idea. I do see potential for an improved UX at the python client. You could have local placeholder variables for the outputs that you can apply other functions to that will then be executed remotely.
You can best view it as a remote session. In terms of UX improvements (also for the UI):

* User only needs to define query and preprocessing steps once
* Waiting time is reduced as queries are only executed once
* Combining your own data adapter with existing algorithms is now possible (even if they are written in different langauges)

> Another thing, will the central algorithms still only be able to execute partial functions that are within the same image? I think this will still hold but I want to check. I think it is a good restriction to have.
The idea is to let this restriction go. And I think we can as:

* Node still controls which images are allowed to run (Nodes should be able to disable the preprocessing by not allowing those images)
* Each image has a specific role as defined in earlier comments (depending on the role then can return something or not)

This makes it predictable what the user can and cant do. Please let me know if you have concerns about this.

> Lastly, just to be sure, I think the main point of this idea is that you don't have to repeat preprocessing steps right? You're not chaining multiple compute tasks one after the other I think.
Indeed that does not make sense to me and is also not possible with the container types we defined: 

Container type | Input type | Output type
-- | -- | --
Data extraction | - | dataframe
Preprocessing | - | dataframe
Algorithm  | dataframe | aggregate
~~Postprocessing~~ | ~~aggregate~~ | ~~aggregate~~

**Comment by fvandaalen** [comment]
Some general concerns;

Currently a major advantage of vante6 is that containers are self-contained processes. The certainty that only the specific process defined in the container is run is a major part of establishing trust between all parties involved.
This is especially advantageous to convince non-technical decisionmakers that what we do is acceptable.

Chainining together containers potentially undercuts this trust as gaps are created in which it is harder to establish trust.
This is especially true when the gaps are explicitly the point, for example when using this approach to preprocess once, store the intermediate result, then run multiple different analysis from this new starting point. It is unclear what can be done with the intermediate result in this "gap".

Providing a framework to show that what happens in these "gaps" can still be trusted will be important.
The most basic solution is of course including a "user beware"-disclaimer; but preferably more is provided to ensure the correct use and provide an easy way to build trust.


In addition to this general concern there is the potential issue of intermediate results containing sensitive information.
This is especially true in the case of vertically split scenarios where preprocessing steps are going to contain information from other sites.

**Comment by frankcorneliusmartin** [comment]
Thank you @fvandaalen for sharing these. We already had a short discussion about this during the community meeting, so these answers are a bit of a repetition of that. 

>  It is unclear what can be done with the intermediate result in this "gap".

I fail to see how this could be exploited. I do see that combinations can be made between parts of code that now are tied together by a single call. The combinations that chainging enables should be considered before approving different images to be run. 

It is possible to make the setup identical to the current implementation: only allow a single container with only single functions for data extraction and computation, in that case you created an identical setup to the current implementation. 

I'll add an collaborations option that only containers from the same image can be chained. If this option is enabled, all types of tasks (extraction, preprocessing, compute, etc.) should be included in the same image (just as the case is now). The difference would be that they would be triggered as different functions.

- [ ] add option to the collaboration to chain only tasks from the same image

> Providing a framework to show that what happens in these "gaps" can still be trusted will be important.
The most basic solution is of course including a "user beware"-disclaimer; but preferably more is provided to ensure the correct use and provide an easy way to build trust.

"Gaps" as you call them are controlled by the process. This data can only be used by the following container to create partials. I also fail to see how someone would be able to exploit these. See how this is controlled in this reply: https://github.com/vantage6/vantage6/issues/943#issuecomment-1938940858

Could you explain maybe a little further how this can be exploited? 

> In addition to this general concern there is the potential issue of intermediate results containing sensitive information.
This is especially true in the case of vertically split scenarios where preprocessing steps are going to contain information from other sites.

In the horizontal case you do not add any more sensitivity to the data that is already there. The `once-per-session` block is intended only for building a local pipeline, and is not able to use any sort of data from other parties.

**Comment by fvandaalen** [comment]
> > It is unclear what can be done with the intermediate result in this "gap".
> 
> I fail to see how this could be exploited. I do see that combinations can be made between parts of code that now are tied together by a single call. The combinations that chainging enables should be considered before approving different images to be run.

One of the best arguments in favour of vantage6 that tends to convince non technical people is that we can tell them the whole analysis is selfcontained. We provide a list of instructions in the form of a docker container and these instructions are the only thing that will be executed. This makes for an easy to understand process where the potential for exploits is limited.

Chaining together multiple containers introduces gaps. This creates 2 potential issues:

1. It becomes possible to chain together containers that should not be chained. This is equivalent to chaining 2 instructions in the same docker container which should not be allowed. This can be controled for by simply checking which containers are chained and properly defining what can and cannot be chained.

2. "Gaps" are created in between the chained containers, and it is unclear what exactly can happen in these gaps.

These gaps create a (theoretical) weakness. First it is an oppertunity for a second (malicious) process to insert itself, second the connections between systems are often the easiest parts to break.
 
> "Gaps" as you call them are controlled by the process. This data can only be used by the following container to create partials. I also fail to see how someone would be able to exploit these. See how this is controlled in this reply: [#943 (comment)](https://github.com/vantage6/vantage6/issues/943#issuecomment-1938940858)
> 
> Could you explain maybe a little further how this can be exploited?

This was not clear during our initial conversation. Limiting the partials to only the follow-up container already takes away some of my concerns.

However, I still have two questions:

1. How do you define a follow up container when there are multiple follow ups? For example, when using 1 preprocessing step to run 3 analyses. How do you stop a 4th analysis from sneaking in?
2. When do you define follow up containers? Does the user define everything at once, or can more follow ups be added later?

If the process works by defining a single super contianer which contains the exact sequence of all subprocesses then my concerns are largerly removed.

 
> > In addition to this general concern there is the potential issue of intermediate results containing sensitive information.
> > This is especially true in the case of vertically split scenarios where preprocessing steps are going to contain information from other sites.
> 
> In the horizontal case you do not add any more sensitivity to the data that is already there. The `once-per-session` block is intended only for building a local pipeline, and is not able to use any sort of data from other parties.

What about a discretization preprocessing step where the bin sizes are dependent on the whole dataset?
E.g. attribute X has the range 0-1 at party A, but the range 3-4 at party B. This is a relativly simple form of preprocessing but already requires the sharing of information to get usefull bins.

**Comment by frankcorneliusmartin** [comment]
> It becomes possible to chain together containers that should not be chained. This is equivalent to chaining 2 instructions in the same docker container which should not be allowed. This can be controled for by simply checking which containers are chained and properly defining what can and cannot be chained.

In a first version there will be not such granular control of which chains are allowed. You either turn it on or completely off so that all steps will be executed from a single image. In a later stadium we could consider this an option. We will add it to the feature list. 

> These gaps create a (theoretical) weakness. First it is an oppertunity for a second (malicious) process to insert itself, second the connections between systems are often the easiest parts to break.

A malicious container could also access the raw data, it does not need the intermediate data frame for this. 

I agree that this is more prone to errors, as we are connecting components. But I do not feel that outweighs the advantages. Breaking is also does not direct lead to a security issue.

> How do you define a follow up container when there are multiple follow ups? For example, when using 1 preprocessing step to run 3 analyses. How do you stop a 4th analysis from sneaking in?

We only define which types of containers are allowed to follow each other up. So a compute step only can only follow up on the output of the `once-per-session`block. This does not limit the amount of compute containers you can send once the session is ready to be used by compute containers.  And so can the `preprocessing` container only follow the `data-extraction` step. 

How would you prevent that in the current infrastructure? There are also no limits on how many analysis you can run? 

> When do you define follow up containers? Does the user define everything at once, or can more follow ups be added later?

The user is allowed to add follow up tasks as long as the session lives. We have not defined end criteria for session yet but these should/could be configured per collaboration and be enforced by the node. We could for example configure it so that it ends when the parent container is done. But these kind of decisions should be up to the collaboration.

> If the process works by defining a single super container which contains the exact sequence of all subprocesses then my concerns are largerly removed.

This is not the case. This could however be an option to configure in your collaboration. In that case session would only be created from parent containers and not from users. This is not something we will implement from the start. 

> What about a discretization preprocessing step where the bin sizes are dependent on the whole dataset?
We do not consider this part of the `once-per-session  block. This should be handled by `compute` containers. The `once-per-session` block can only use the local data.

**Comment by fvandaalen** [comment]
> > These gaps create a (theoretical) weakness. First it is an oppertunity for a second (malicious) process to insert itself, second the connections between systems are often the easiest parts to break.
> 
> A malicious container could also access the raw data, it does not need the intermediate data frame for this.
> 
> I agree that this is more prone to errors, as we are connecting components. But I do not feel that outweighs the advantages. Breaking it also does not direct lead to a security issue.

True, however I'm playing devil's advocate and trying to imagine the worst possible scenarios.
A malicious container may be just as dangerous, but a malicious container is easier to detect as images can be vetted by experts belonging to every party. This allows each party to independently establish trust in the process before any analysis is executed. However, chained containers are considerable harder to vet as it is possible for a researcher to define them on the fly (if I understand it correctly) this makes it harder to establish that same level of trust. This is ultimatly also my main concern; that it will undercut the level of trust that vantage6 currently brings.
 
> > How do you define a follow up container when there are multiple follow ups? For example, when using 1 preprocessing step to run 3 analyses. How do you stop a 4th analysis from sneaking in?
> 
> We only define which types of containers are allowed to follow each other up. So a compute step only can only follow up on the output of the `once-per-session`block. This does not limit the amount of compute containers you can send once the session is ready to be used by compute containers. And so can the `preprocessing` container only follow the `data-extraction` step.
> 
> How would you prevent that in the current infrastructure? There are also no limits on how many analysis you can run?
> 
> > When do you define follow up containers? Does the user define everything at once, or can more follow ups be added later?
> 
> The user is allowed to add follow up tasks as long as the session lives. We have not defined end criteria for session yet but these should/could be configured per collaboration and be enforced by the node. We could for example configure it so that it ends when the parent container is done. But these kind of decisions should be up to the collaboration.

Allowing users to add follow up tasks as long as the session lives would definitly be dangerous and undercuts the containerization principles I've mentioned before.
Is it possible to specify which specific follow up tasks are allowed?

By which I mean is it possible to easily define something like 
- preprocessing with image "X" may only be followed by a computer container with image "Y", it may never be followed by a compute container with image "Z"?
And have this definition live in a place that is easily auditable by every relevant party.

This can be in a sort of super container, but I would also accept a centrally arranged whitelisting for the collaboration.

> > If the process works by defining a single super container which contains the exact sequence of all subprocesses then my concerns are largerly removed.
> 
> This is not the case. This could however be an option to configure in your collaboration. In that case session would only be created from parent containers and not from users. This is not something we will implement from the start.

I think that this is ultimatly the solution you will need to establish that the system as a whole remains trustworthy by continuing to limit the direct influence a researcher has on any and all processes within vantage6.


> > What about a discretization preprocessing step where the bin sizes are dependent on the whole dataset?
> 
> We do not consider this part of the `once-per-session` block. This should be handled by `compute` containers. The `once-per-session` block can only use the local data.

That does limit the potential problems, but also limits the potential utility.
If the preprocessing steps are 100% limited to fully local calculations then my concerns are largely moot.
However, in this case I would also argue that the preprocessing could just be done locally outside of vantage6.
There is still utility in executing this within vantage6 in certain scenarios; but it is far more limited than what I initially imagined when hearing the idea.

**Comment by frankcorneliusmartin** [comment]
> True, however I'm playing devil's advocate and trying to imagine the worst possible scenarios.

I am happy that you do. And if we can improve the overal security of the infrastructure with this feature I think we should do it. However this is not a security feature, so keeping security on par is acceptable. 

> A malicious container may be just as dangerous, but a malicious container is easier to detect as images can be vetted by experts belonging to every party. This allows each party to independently establish trust in the process before any analysis is executed. However, chained containers are considerable harder to vet as it is possible for a researcher to define them on the fly (if I understand it correctly) this makes it harder to establish that same level of trust. This is ultimatly also my main concern; that it will undercut the level of trust that vantage6 currently brings.

So to counter this:

* You still are able to limit the chaining to a single container. And within this container you define only 2 functions, a `data_extraction` and a `compute` function. Then the parties have to deal with the same amount of code and no alternative combinations are possible. So I would argue that in this case you can have an identical level of trust as in the current implementation.
* I would argue that this system could improve trust too. 
  * Once we have solid `data_extraction` and `preprocessing` containers used and reviewed by many (documented in the algorithm store) this should improve trust in these components. In the current system you need to inject these bits of code for every image, making that you have to review more code and that you easier slip in a malicious line. 
  * Each function now has a more define purpose making it easier to spot that it does something that should not be within its role.  
  * Each container type gets only what it needs. Thus a data extraction container gets access to the data, but for example cannot reach the vantage6-server. The compute container can reach the intermediate dataframe but not the data source directly.

> Allowing users to add follow up tasks as long as the session lives would definitly be dangerous and undercuts the containerization principles I've mentioned before.

This is only the case when you give the users some flexibility with choosing different containers to succeed each other, do you agree? Or is this also a problem if we only allow Image X which has only 2 dedicated functions? If you agree that you should set the `session_restrict_to_same_image` to True and then X can only follow X. 

We could work on some settable policies to determine how long a session lives, for example:

* It could be a consideration for a collaboration that you only let your session live as long as the parent container lives. 
* For x days/hours/minutes
* User deleted
* After X runs on the intermediate

> By which I mean is it possible to easily define something like
preprocessing with image "X" may only be followed by a computer container with image "Y", it may never be followed by a compute container with image "Z"?
And have this definition live in a place that is easily auditable by every relevant party.
This can be in a sort of super container, but I would also accept a centrally arranged whitelisting for the collaboration.

I think the requirement of a central/super container should be optional as this would limit the potential of chaining. As this would users not be able to reuse sessions at all. Think about the following scenarios where this could be useful:

* A user needs to compute a GLM and a CoxPH model on the same preprocessed data from an OMOP CDM. The query takes about 10 minutes to execute for every iteration. 
* A user is not able to create sessions but relies on a collaboration manager to define the session for him. The collaboration manager can define the query part and the preprocessing. For example he could add a differential privacy step after the extraction. 
* A user want to share a session with another user

Note that all these permissions are within the vantage6-permission system. Thus not every user can create sessions and not every user can create every scope of a session (own, org, col). 

I think the 'flow'-whitelisting idea could be implemented which could be a nice additional security feature, so lets create a separate issue for that. For now, to keep the security on par you could use the `session_restrict_to_same_image` option that is tied to the collaboration (and enforced at the nodes). This is basically your super container scenario. 

> I think that this is ultimatly the solution you will need to establish that the system as a whole remains trustworthy by continuing to limit the direct influence a researcher has on any and all processes within vantage6.

I think this is up to the collaboration to decide. Its a security/privacy vs flexibility discussion. The infra should provide the tools and options to support both cases.

> However, in this case I would also argue that the preprocessing could just be done locally outside of vantage6.

Partially true. However this allows user to have some level of control over the data_extraction / preprocessing step depending on the flexibility that these containers offer. Some examples of preprocessing interfaces that **could** be opened up to the user:

* Filter the data
* Apply data transformations
* Apply one hot encoding
* Stratify data

The data_extraction step is mostly intereseting when looking at 'external data sources' (=non file based), so for example if we need to query an SQL instance we need an SQL statement. The user could have some control over this statement (or complete, if the collaboration would allow that. I would not recommend that). 

The intermediate dataframe also makes sure that there is no distinction between file based and non-file based databases. This helps reporting for example the variables (column names) to the user and show these as a drop down in the UI. This also allows us to re-use the CoxPH algorithm without modification for any data source that we have a `data_extraction` container for. 

> There is still utility in executing this within vantage6 in certain scenarios; but it is far more limited than what I initially imagined when hearing the idea.

Feel free to create an additional issue for this. But as you mentioned, and I agree, this poses bigger problems as you would locally store results that are coming from partials from other data parties. However I **do** think we could support such cases, and maybe leave it up to the collaboration to decide whether this is acceptable or not.

**Comment by fvandaalen** [comment]
> > True, however I'm playing devil's advocate and trying to imagine the worst possible scenarios.
> 
> I am happy that you do. And if we can improve the overal security of the infrastructure with this feature I think we should do it. However this is not a security feature, so keeping security on par is acceptable.
> 
> > A malicious container may be just as dangerous, but a malicious container is easier to detect as images can be vetted by experts belonging to every party. This allows each party to independently establish trust in the process before any analysis is executed. However, chained containers are considerable harder to vet as it is possible for a researcher to define them on the fly (if I understand it correctly) this makes it harder to establish that same level of trust. This is ultimatly also my main concern; that it will undercut the level of trust that vantage6 currently brings.
> 
> So to counter this:
> 
>     * You still are able to limit the chaining to a single container. And within this container you define only 2 functions, a `data_extraction` and a `compute` function. Then the parties have to deal with the same amount of code and no alternative combinations are possible. So I would argue that in this case you can have an identical level of trust as in the current implementation.
> 
Sure, you can always stick with the old approach. But I am wondering to which extend people are going to unwilling to use the new feature due to the lower level of trust.

>     * I would argue that this system could improve trust too.
>       
>       * Once we have solid `data_extraction` and `preprocessing` containers used and reviewed by many (documented in the algorithm store) this should improve trust in these components. In the current system you need to inject these bits of code for every image, making that you have to review more code and that you easier slip in a malicious line.
>       * Each function now has a more define purpose making it easier to spot that it does something that should not be within its role.
>       * Each container type gets only what it needs. Thus a data extraction container gets access to the data, but for example cannot reach the vantage6-server. The compute container can reach the intermediate dataframe but not the data source directly.
> 

There are definite advantages, especially with respect to ease of use and seperation of functionality.
But it also gives the end-user more control as it is difficult to control which images get chained together. 
In my experience the biggest problem with convincing lawyers & managers is that they are afraid they will lose control of their data.
Hence why I'm making such a big deal out of it.

> > Allowing users to add follow up tasks as long as the session lives would definitly be dangerous and undercuts the containerization principles I've mentioned before.
> 
> This is only the case when you give the users some flexibility with choosing different containers to succeed each other, do you agree? Or is this also a problem if we only allow Image X which has only 2 dedicated functions? If you agree that you should set the `session_restrict_to_same_image` to True and then X can only follow X.

I'd argue that chaining is only going to be usefull if you can chain distinct images. If you would want to chain the same image you might as well merge the two dedicated functions (or write a third function that calls both as subfunctions)

Sure this gives the same level of security as vantage6 currently has. But it also makes the new feature kind of redundant.
 
> I think the 'flow'-whitelisting idea could be implemented which could be a nice additional security feature, so lets create a separate issue for that. For now, to keep the security on par you could use the `session_restrict_to_same_image` option that is tied to the collaboration (and enforced at the nodes). This is basically your super container scenario.
> 
> > I think that this is ultimatly the solution you will need to establish that the system as a whole remains trustworthy by continuing to limit the direct influence a researcher has on any and all processes within vantage6.
> 
> I think this is up to the collaboration to decide. Its a security/privacy vs flexibility discussion. The infra should provide the tools and options to support both cases.

I think that 'flow'-whitelisting should be the default behaviour.
It allows the various parties to retain control over what processes can be run in what order. Which will please the lawyers & managers.
It gives the flexibility to reuse and chain together existing images, bringing the full advantage of chaining with it.
And if a collaboration is feeling particularly adventurous they can always just whitelist everything.


> 
> > However, in this case I would also argue that the preprocessing could just be done locally outside of vantage6.
> 
> Partially true. However this allows user to have some level of control over the data_extraction / preprocessing step depending on the flexibility that these containers offer. Some examples of preprocessing interfaces that **could** be opened up to the user:
> 
>     * Filter the data
> 
>     * Apply data transformations
> 
>     * Apply one hot encoding
> 
>     * Stratify data
> 
> 
> The data_extraction step is mostly intereseting when looking at 'external data sources' (=non file based), so for example if we need to query an SQL instance we need an SQL statement. The user could have some control over this statement (or complete, if the collaboration would allow that. I would not recommend that).
> 
> The intermediate dataframe also makes sure that there is no distinction between file based and non-file based databases. This helps reporting for example the variables (column names) to the user and show these as a drop down in the UI. This also allows us to re-use the CoxPH algorithm without modification for any data source that we have a `data_extraction` container for.

Quick question; how will this work with non-python based containers?

> > There is still utility in executing this within vantage6 in certain scenarios; but it is far more limited than what I initially imagined when hearing the idea.
> 
> Feel free to create an additional issue for this. But as you mentioned, and I agree, this poses bigger problems as you would locally store results that are coming from partials from other data parties. However I **do** think we could support such cases, and maybe leave it up to the collaboration to decide whether this is acceptable or not.

'Flow'-whitelisting would probably cover most of my immeadiate concerns while allowing most of the functionality I had in mind.
The concerns that remain regarding partial results when using flow-whitelisting would be similar to the concerns that arise when using partials as part of a single container.

**Comment by frankcorneliusmartin** [comment]
> I'd argue that chaining is only going to be useful if you can chain distinct images. If you would want to chain the same image you might as well merge the two dedicated functions (or write a third function that calls both as subfunctions)
Sure this gives the same level of security as vantage6 currently has. But it also makes the new feature kind of redundant.

Not true, as: 
* this would give a unified way of working. No longer dependent on the type of data source. This means that the UX for the end user would be identical whether he or she is using a CSV, SPARQL or whatever.
* Expensive queries only have to be executed once. For example if we use OMOP CDM with an iterative algorithm and each iteration would take 10m to query (which is the case) the analysis could easily take up hours. While if we can safe the intermediate dataframe this could be handled in 10m querying + 3 min for the analysis.

> I think that 'flow'-whitelisting should be the default behaviour.
It allows the various parties to retain control over what processes can be run in what order. Which will please the lawyers & managers.
It gives the flexibility to reuse and chain together existing images, bringing the full advantage of chaining with it.
And if a collaboration is feeling particularly adventurous they can always just whitelist everything.

I agree that this is a desirable option and should be implemented. But since we can get the same security as we currently have when the option `session_restrict_to_same_image` I do not see that it is unsafe to release this feature without it. I be happy to incorporate the work of others on this feature, however its just not something that I require for any of my projects in the current timeline. Maybe this is something required in the STRONG-AYA project, but that is something to discuss with @bartvanb. 

> Quick question; how will this work with non-python based containers?
From the infrastructure perspective all containers will be mounted with the same files / environment vars regardless of their language. That said, the vantage6-algorithm-tools provide convenient functions to handle this IO. If you write any of the components in a different language you are required to follow the vantage6 standard. 

This is also an advantage of chaining, it is possible to combine different parts that are written in completely different langauges. For example you can use a python preprocessing algorithm and the R CoxPH version.

> 'Flow'-whitelisting would probably cover most of my immeadiate concerns while allowing most of the functionality I had in mind.
The concerns that remain regarding partial results when using flow-whitelisting would be similar to the concerns that arise when using partials as part of a single container.

Ok noted. We will (really) consider releasing it with this feature at the same time. I would have to investigate a little how much effort it would require.

**Comment by fvandaalen** [comment]
> > I'd argue that chaining is only going to be useful if you can chain distinct images. If you would want to chain the same image you might as well merge the two dedicated functions (or write a third function that calls both as subfunctions)
> > Sure this gives the same level of security as vantage6 currently has. But it also makes the new feature kind of redundant.
> 
> Not true, as:
> 
>     * this would give a unified way of working. No longer dependent on the type of data source. This means that the UX for the end user would be identical whether he or she is using a CSV, SPARQL or whatever.
>     * Expensive queries only have to be executed once. For example if we use OMOP CDM with an iterative algorithm and each iteration would take 10m to query (which is the case) the analysis could easily take up hours. While if we can safe the intermediate dataframe this could be handled in 10m querying + 3 min for the analysis.

True, however since federated learning isn't exactly known for its high speed & Vantage6 is largely focused on scenarios where we can usually afford to wait quite some time for an analysis to finish I personally would not value this particularly highly.

> > I think that 'flow'-whitelisting should be the default behaviour.
> > It allows the various parties to retain control over what processes can be run in what order. Which will please the lawyers & managers.
> > It gives the flexibility to reuse and chain together existing images, bringing the full advantage of chaining with it.
> > And if a collaboration is feeling particularly adventurous they can always just whitelist everything.
> 
> I agree that this is a desirable option and should be implemented. But since we can get the same security as we currently have when the option `session_restrict_to_same_image` I do not see that it is unsafe to release this feature without it. I be happy to incorporate the work of others on this feature, however its just not something that I require for any of my projects in the current timeline. Maybe this is something required in the STRONG-AYA project, but that is something to discuss with @bartvanb.

Fair enough. However, I do think the 'flow'-whitelisting will be necesary for chaining to see widespread adoption.
I am personally most interested in chaining together containers that are not purely local (e.g. things like feature selection, organisation-wide discretization).
And those will it. 

> > Quick question; how will this work with non-python based containers?
> 
> From the infrastructure perspective all containers will be mounted with the same files / environment vars regardless of their language. That said, the vantage6-algorithm-tools provide convenient functions to handle this IO. If you write any of the components in a different language you are required to follow the vantage6 standard.
> 
> This is also an advantage of chaining, it is possible to combine different parts that are written in completely different langauges. For example you can use a python preprocessing algorithm and the R CoxPH version.

You mentioned earlier that for example a compute container cannot reach the original raw data and would only be able to reach the dataframe left behind by the data_extraction container.

I meant how does that work if you use different languages? For example, Java is going to struggle to open that dataframe.

> > 'Flow'-whitelisting would probably cover most of my immeadiate concerns while allowing most of the functionality I had in mind.
> > The concerns that remain regarding partial results when using flow-whitelisting would be similar to the concerns that arise when using partials as part of a single container.
> 
> Ok noted. We will (really) consider releasing it with this feature at the same time. I would have to investigate a little how much effort it would require.
Great!

**Comment by frankcorneliusmartin** [comment]
> True, however since federated learning isn't exactly known for its high speed & Vantage6 is largely focused on scenarios where we can usually afford to wait quite some time for an analysis to finish I personally would not value this particularly highly.

There are more advantages. Not all features are useful for everyone, see them listed at the top: https://github.com/vantage6/vantage6/issues/943#issue-2026208385

> Fair enough. However, I do think the 'flow'-whitelisting will be necesary for chaining to see widespread adoption.
I am personally most interested in chaining together containers that are not purely local (e.g. things like feature selection, organisation-wide discretization).
And those will it.

You still can implement these at algorithm level for now. 


> You mentioned earlier that for example a compute container cannot reach the original raw data and would only be able to reach the dataframe left behind by the data_extraction container.

> I meant how does that work if you use different languages? For example, Java is going to struggle to open that dataframe.
data frames exported from `data-extraction` or `pre-processing` are stored in [parquet ](https://parquet.apache.org/) format. There are client libraries available for all major languages, for java specifically: https://github.com/apache/parquet-mr

**Comment by fvandaalen** [comment]
> > You mentioned earlier that for example a compute container cannot reach the original raw data and would only be able to reach the dataframe left behind by the data_extraction container.
> 
> > I meant how does that work if you use different languages? For example, Java is going to struggle to open that dataframe.
> 
> Data frames exported from `data-extraction` or `pre-processing` are stored in [parquet ](https://parquet.apache.org/) format. There are client libraries available for all major languages, for java specifically: https://github.com/apache/parquet-mr
Cool, just doublechecking. Was scared for a moment that it would've been something weird, but parquet I've used.

**Comment by bartvanb** [comment]
Basic functionality merged, rest of work to do is in #1495
